File: frontend/src/services/host.ts
import api from '../config/api';

export interface Host {
  id: number;
  project_id?: number;
  hostname: string;
  ip_address: string;
  salt_minion_id?: string;
  os_type?: string;
  os_version?: string;
  cpu_cores?: number;
  memory_gb?: number;
  disk_gb?: number;
  status: 'online' | 'offline' | 'unknown';
  environment?: string;
  cloud_platform_id?: number;
  ssh_port?: number;
  ssh_username?: string;
  ssh_key_id?: number;
  last_seen_at?: string;
  salt_version?: string;
  description?: string;
  created_at?: string;
  updated_at?: string;
  project?: {
    id: number;
    name: string;
  };
  cloud_platform?: {
    id: number;
    name: string;
    display_name: string;
    color?: string;
  };
  groups?: Array<{ id: number; name: string }>;
  tags?: Array<{ id: number; name: string; color?: string }>;
}

export interface CreateHostRequest {
  project_id: number;
  hostname: string;
  ip_address: string;
  salt_minion_id?: string;
  os_type?: string;
  os_version?: string;
  cpu_cores?: number;
  memory_gb?: number;
  disk_gb?: number;
  status?: string;
  environment?: string;
  cloud_platform_id?: number;
  ssh_port?: number;
  ssh_username?: string;
  ssh_key_id?: number;
  description?: string;
}

export interface ListHostsResponse {
  hosts: Host[];
  total: number;
  page: number;
  page_size: number;
}

export interface HostFilters {
  project_id?: number;
  hostname?: string;
  ip_address?: string;
  status?: string;
  environment?: string;
  cloud_platform_id?: number;
  group_id?: number;
  tag_id?: number;
}

export const hostService = {
  list: async (
    page: number = 1,
    pageSize: number = 10,
    filters?: HostFilters
  ): Promise<ListHostsResponse> => {
    const response = await api.get('/hosts', {
      params: { page, page_size: pageSize, ...filters },
    });
    return response.data;
  },

  get: async (id: number): Promise<{ host: Host }> => {
    const response = await api.get(`/hosts/${id}`);
    return response.data;
  },

  create: async (data: CreateHostRequest): Promise<{ host: Host }> => {
    const response = await api.post('/hosts', data);
    return response.data;
  },

  update: async (id: number, data: Partial<CreateHostRequest>): Promise<void> => {
    await api.put(`/hosts/${id}`, data);
  },

  delete: async (id: number): Promise<void> => {
    await api.delete(`/hosts/${id}`);
  },

  addToGroup: async (hostId: number, groupId: number): Promise<void> => {
    await api.post(`/hosts/${hostId}/groups`, { group_id: groupId });
  },

  removeFromGroup: async (hostId: number, groupId: number): Promise<void> => {
    await api.delete(`/hosts/${hostId}/groups`, {
      data: { group_id: groupId },
    });
  },

  addTag: async (hostId: number, tagId: number): Promise<void> => {
    await api.post(`/hosts/${hostId}/tags`, { tag_id: tagId });
  },

  removeTag: async (hostId: number, tagId: number): Promise<void> => {
    await api.delete(`/hosts/${hostId}/tags`, {
      data: { tag_id: tagId },
    });
  },

  // 同步主机状态（通过 Salt）
  syncStatus: async (hostId: number): Promise<{ host: Host; message: string }> => {
    const response = await api.post(`/hosts/${hostId}/sync-status`);
    return response.data;
  },

  // 同步所有主机状态
  syncAllStatus: async (): Promise<{ message: string }> => {
    const response = await api.post('/hosts/sync-all-status');
    return response.data;
  },

  // 发现所有 Salt Minion
  discoverMinions: async (): Promise<{ minions: MinionInfo[]; total: number }> => {
    const response = await api.get('/hosts/discover-minions');
    return response.data;
  },

  // 自动为主机分配 Salt Minion ID
  autoAssignMinion: async (hostId: number): Promise<{ host: Host; minion_id: string | null; message: string }> => {
    const response = await api.post(`/hosts/${hostId}/auto-assign-minion`);
    return response.data;
  },
};

// Salt Minion 信息
export interface MinionInfo {
  id: string;
  ip_address: string;
  hostname: string;
  os: string;
  os_release: string;
}

File: backend/internal/models/host.go
package models

import (
	"time"

	"gorm.io/gorm"
)

// Host 主机模型
type Host struct {
	ID           uint64         `gorm:"primaryKey" json:"id"`
	ProjectID    uint64         `gorm:"not null;index" json:"project_id"`
	Hostname     string         `gorm:"size:255;not null;index" json:"hostname"`
	IPAddress    string         `gorm:"size:45;not null;index" json:"ip_address"`
	SaltMinionID string         `gorm:"uniqueIndex;size:255" json:"salt_minion_id"`
	OSType       string         `gorm:"size:50" json:"os_type"`
	OSVersion    string         `gorm:"size:100" json:"os_version"`
	CPUCores     *int           `json:"cpu_cores"`
	MemoryGB     *float64       `gorm:"type:decimal(10,2)" json:"memory_gb"`
	DiskGB       *float64       `gorm:"type:decimal(10,2)" json:"disk_gb"`
	Status       string         `gorm:"size:20;not null;default:'unknown';index" json:"status"`
	Environment  string         `gorm:"size:20;index" json:"environment"`
	SSHPort      int            `gorm:"not null;default:22" json:"ssh_port"`
	SSHKeyID     *uint64        `gorm:"index" json:"ssh_key_id"` // 可选的默认SSH密钥ID
	LastSeenAt   *time.Time     `json:"last_seen_at"`
	SaltVersion  string         `gorm:"size:50" json:"salt_version"`
	Metadata     string         `gorm:"type:jsonb" json:"metadata"`
	Description  string         `json:"description"`
	CreatedAt    time.Time      `json:"created_at"`
	UpdatedAt    time.Time      `json:"updated_at"`
	DeletedAt    gorm.DeletedAt `gorm:"index" json:"-"`

	// 关联
	Project *Project    `gorm:"foreignKey:ProjectID" json:"project,omitempty"`
	SSHKey  *SSHKey     `gorm:"foreignKey:SSHKeyID" json:"ssh_key,omitempty"`
	Groups  []HostGroup `gorm:"many2many:host_group_members;" json:"groups,omitempty"`
	Tags    []HostTag   `gorm:"many2many:host_tag_assignments;" json:"tags,omitempty"`
}

// TableName 指定表名
func (Host) TableName() string {
	return "hosts"
}

// HostGroup 主机组模型
type HostGroup struct {
	ID          uint64         `gorm:"primaryKey" json:"id"`
	ProjectID   uint64         `gorm:"not null;index" json:"project_id"`
	Name        string         `gorm:"size:100;not null;index" json:"name"`
	Description string         `json:"description"`
	CreatedAt   time.Time      `json:"created_at"`
	UpdatedAt   time.Time      `json:"updated_at"`
	DeletedAt   gorm.DeletedAt `gorm:"index" json:"-"`

	// 关联
	Project *Project `gorm:"foreignKey:ProjectID" json:"project,omitempty"`
	Hosts   []Host   `gorm:"many2many:host_group_members;" json:"hosts,omitempty"`
}

// TableName 指定表名
func (HostGroup) TableName() string {
	return "host_groups"
}

// HostTag 主机标签模型
type HostTag struct {
	ID          uint64    `gorm:"primaryKey" json:"id"`
	Name        string    `gorm:"uniqueIndex;size:50;not null" json:"name"`
	Color       string    `gorm:"size:7" json:"color"`
	Description string    `json:"description"`
	CreatedAt   time.Time `json:"created_at"`

	// 关联
	Hosts []Host `gorm:"many2many:host_tag_assignments;" json:"hosts,omitempty"`
}

// TableName 指定表名
func (HostTag) TableName() string {
	return "host_tags"
}

// HostGroupMember 主机组成员关联模型
type HostGroupMember struct {
	ID        uint64    `gorm:"primaryKey"`
	HostID    uint64    `gorm:"not null;index"`
	GroupID   uint64    `gorm:"not null;index"`
	CreatedAt time.Time `gorm:"not null;default:now()"`
}

// TableName 指定表名
func (HostGroupMember) TableName() string {
	return "host_group_members"
}

// HostTagAssignment 主机标签关联模型
type HostTagAssignment struct {
	ID        uint64    `gorm:"primaryKey"`
	HostID    uint64    `gorm:"not null;index"`
	TagID     uint64    `gorm:"not null;index"`
	CreatedAt time.Time `gorm:"not null;default:now()"`
}

// TableName 指定表名
func (HostTagAssignment) TableName() string {
	return "host_tag_assignments"
}
